File: rust/herb-linter/src/rules/erb_no_sleep.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::prism_utils::{is_sleep_call, walk_prism};
use crate::utils::source_slice::location_from_offset;

use herb::ParseResult;
use herb_config::{Severity, SeverityConfig};

pub struct ERBNoSleepRule;

impl Rule for ERBNoSleepRule {
  fn name(&self) -> &'static str {
    "erb-no-sleep"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("unreleased")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::PerMode {
      cli: Severity::Error,
      editor: Severity::Info,
    }
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      prism_program: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

impl ParserRule for ERBNoSleepRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    let prism_node = match result.value.prism() {
      Some(prism_node) => prism_node,
      None => return Vec::new(),
    };

    let source = if context.source.is_empty() { &result.source } else { &context.source };

    if source.is_empty() {
      return Vec::new();
    }

    let mut offenses = Vec::new();

    walk_prism(prism_node, &mut |node| {
      if is_sleep_call(node) {
        let call_source = source.get(node.start_offset..node.end_offset).unwrap_or("");

        offenses.push(UnboundOffense::new(
          self.name(),
          format!(
            "Avoid using `{}` in ERB templates. It blocks the thread rendering the response and delays the page for every request. Remove it, or move the delay outside of the template.",
            call_source
          ),
          location_from_offset(source, node.start_offset, node.end_offset),
        ));
      }

      true
    });

    offenses
  }
}

File: javascript/packages/linter/src/rules/erb-no-sleep.ts
import { ParserRule } from "../types.js"
import { PrismVisitor, substringFromByteOffset , locationFromByteOffset } from "@herb-tools/core"

import { isSleepCall } from "./prism-rule-utils.js"

import type { ParseResult, ParserOptions, PrismNode } from "@herb-tools/core"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"

class SleepCallCollector extends PrismVisitor {
  public readonly calls: PrismNode[] = []

  visitCallNode(node: PrismNode): void {
    if (isSleepCall(node)) {
      this.calls.push(node)
    }

    this.visitChildNodes(node)
  }
}

export class ERBNoSleepRule extends ParserRule {
  static ruleName = "erb-no-sleep"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: {
        cli: "error",
        editor: "info",
      },
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_program: true,
    }
  }

  check(result: ParseResult, _context?: Partial<LintContext>): UnboundLintOffense[] {
    const source = result.value.source
    const prismNode = result.value.prismNode

    if (!prismNode || !source) return []

    const collector = new SleepCallCollector()
    collector.visit(prismNode)

    return collector.calls.map(call => {
      const { startOffset, length } = call.location
      const location = locationFromByteOffset(source, startOffset, length)
      const callSource = substringFromByteOffset(source, startOffset, length)

      return this.createOffense(
        `Avoid using \`${callSource}\` in ERB templates. It blocks the thread rendering the response and delays the page for every request. Remove it, or move the delay outside of the template.`,
        location,
      )
    })
  }
}
